Automattic/wp-calypso
# Trail Map: Provision for rendering grouped features in the Features Grid

Related to https://github.com/Automattic/wp-calypso/issues/89682
Based on https://github.com/Automattic/wp-calypso/pull/89840

## Proposed Changes

Implements the desktop and mobile (organically) functionalities for rendering grouped features in the Features Grid when the associated props are set. 

- This should handle the core functionality, with the right order shown in the Figma mocks. 
    - Styling fixes and some data restructuring will need to be done in a separate PR (see "Next/TODO" below).
- The data changes follow a similar route to the Comparison Grid - feature groups to be defined in `calypso-products`, each referring to a list of features and the plans having their associated flat list of features derived from the existing selectors
- The properties `renderCategorisedFeatures` and `featureGroupMap` are defined to control the behavior (setting `renderCategorisedFeatures` to `false` will work exactly like before.
    - These are set according to the `useTrailMapExperiment` hook
- No changes to the comparison grid, just pulling out the `featureGroupMap`. Comparison Grid should work like before.

This PR can be merged. Ideally, we will not need to touch much else on the grid components outside of styling fixes. The rest is down to feeding the right data. See Next/TODO below.

## Media

### No categories

<img width="600" alt="Screenshot 2024-04-26 at 1 41 56 PM" src="https://github.com/Automattic/wp-calypso/assets/1705499/6cacbe1b-3616-46e3-9f69-c1334dbd6630">

### With categories

<img width="600" alt="Screenshot 2024-04-26 at 1 41 25 PM" src="https://github.com/Automattic/wp-calypso/assets/1705499/412b730c-df1b-43d3-8d2f-496a2f1cd94a">

## Next/TODO

### Styling fixes
- Feature group titles
- Spacing
- Not sure what else

### Data structures

We need to provide the full list of features for each plan, not just the new ones. This will need to be done on condition, since the current data flows expect that each plan carries its new/unique features. We can do this in a couple of ways. Including (others possible too): 

1. Provide the full list of features for each plan from the plans-list selectors in `calypso-products`. This could be preferred if we want to reflect reality directly at the core data layer (and we can do something like call the previous plan's selector if we want to reduce redundancy). However, we will likely need a bigger cleanup in the end when the experiment flags are removed. We will also need a **likely small** refactor (now) in `use-restructured-plan-features-for-comparison-grid` to not re-compile the lists with `...previousPlanFeatures`.
2. Create/update the logic in `use-plan-features-for-grid-plans` to recompile the features like `use-restructured-plan-features-for-comparison-grid` does. This could be the fastest, as we will likely only need to pass a prop into the hook and copy/paste the same code across. The selectors from `calypso-products` will continue to provide the new/unique features for each plan.

So on the above, we can do `2` for now if it's fastest, and do `1` in the cleanup phase. 

## Testing Instructions



- Go to `/start/plans?flags=onboarding/trail-map-feature-grid` and confirm the media above. 
    - Features should be grouped and reordered to match the designs.
    - Confirm tablet/mobile that everything functions correctly
    - Confirm the table borders top to bottom that they are correctly rendered
- Go to `/start/plans?flags=-onboarding/trail-map-feature-grid` (flag disabled) and confirm the grid renders exactly like before. 

## Pre-merge Checklist
